File: src/faker_mapper.py
"""
Consistent PII Synthetic Replacement Mapper.
Ensures that the same real PII entity is always replaced by the exact same synthetic substitute.
For example:
"Rashi Patil" -> "John Doe"
"rashhi.patil@gmail.com" -> "john.doe@example.com"
"+91 9876543210" -> "+91 1234567645"
"""

import hashlib
import random
from typing import Dict
try:
    from faker import Faker
    fake = Faker()
except ImportError:
    fake = None


class SyntheticPIIMapper:
    """
    Maintains a deterministic map from original PII text to synthetic alternatives.
    """

    def __init__(self, seed: int = 42):
        self.seed = seed
        self.mapping: Dict[str, str] = {}
        if fake:
            Faker.seed(seed)
        
        # Pre-seeded specific examples to mirror sample assignment behavior
        self._preset_mappings = {
            "Rashi Patil": "John Doe",
            "rashhi.patil@gmail.com": "john.doe@example.com",
            "Rohan Dey": "Peter Parker",
            "rohan.dey@gmail.com": "peter.parker@example.com",
            "+91 9876543210": "+91 1234567645"
        }

    def _get_deterministic_int(self, text: str) -> int:
        h = hashlib.sha256(text.encode('utf-8')).hexdigest()
        return int(h[:8], 16)

    def get_replacement(self, original_text: str, pii_type: str) -> str:
        """
        Retrieves or generates a consistent fake substitute for original_text.
        """
        clean_text = original_text.strip()
        if not clean_text:
            return original_text

        if clean_text in self._preset_mappings:
            return self._preset_mappings[clean_text]

        if clean_text in self.mapping:
            return self.mapping[clean_text]

        seed_val = self._get_deterministic_int(clean_text)
        rng = random.Random(self.seed + seed_val)

        if pii_type == "FULL_NAME":
            first_names = ["Alex", "Jordan", "Morgan", "Taylor", "Casey", "Riley", "Sam", "Chris", "Pat", "Drew", "David", "Michael", "Sarah", "Emily"]
            last_names = ["Smith", "Johnson", "Williams", "Brown", "Jones", "Garcia", "Miller", "Davis", "Rodriguez", "Martinez"]
            fake_name = f"{rng.choice(first_names)} {rng.choice(last_names)}"
            replacement = fake_name

        elif pii_type == "EMAIL":
            user_part = clean_text.split("@")[0].lower()
            # replace non-alphanumeric
            clean_user = "".join(c for c in user_part if c.isalnum() or c in ".")
            replacement = f"{clean_user}.fake@example.com"

        elif pii_type == "PHONE_NUMBER":
            # Preserve prefix if present e.g. +91
            digits = "".join(rng.choices("0123456789", k=10))
            if clean_text.startswith("+91"):
                replacement = f"+91 {digits}"
            elif clean_text.startswith("+"):
                prefix = clean_text.split()[0]
                replacement = f"{prefix} {digits}"
            else:
                replacement = f"+91 {digits[:5]} {digits[5:]}"

        elif pii_type == "COMPANY_NAME":
            fake_comps = ["Acme Corp", "Apex Technologies", "Global Industries", "Omni Enterprises", "Zenith Holdings", "Nexus Solutions"]
            replacement = rng.choice(fake_comps)

        elif pii_type == "ADDRESS":
            streets = ["123 Innovation Way", "456 Tech Park Boulevard", "789 Corporate Plaza", "101 Enterprise Avenue"]
            cities = ["Metro City", "Techville", "Innovate Town", "Cyber City"]
            replacement = f"{rng.choice(streets)}, {rng.choice(cities)} - 400001, Maharashtra, India"

        elif pii_type == "GOVT_ID":
            # PAN / DIN / SSN
            if len(clean_text) == 10 and clean_text[:5].isalpha(): # PAN format
                letters = "".join(rng.choices("ABCDEFGHIJKLMNOPQRSTUVWXYZ", k=5))
                nums = "".join(rng.choices("0123456789", k=4))
                last_char = rng.choice("ABCDEFGHIJKLMNOPQRSTUVWXYZ")
                replacement = f"{letters}{nums}{last_char}"
            elif "DIN" in clean_text.upper() or len(clean_text) == 8:
                replacement = "DIN: " + "".join(rng.choices("0123456789", k=8))
            elif "-" in clean_text and len(clean_text) == 11: # SSN
                replacement = f"{rng.randint(100,999):03d}-{rng.randint(10,99):02d}-{rng.randint(1000,9999):04d}"
            else:
                replacement = "ID-" + "".join(rng.choices("0123456789", k=9))

        elif pii_type == "CREDIT_CARD":
            replacement = "4111-XXXX-XXXX-1111"

        elif pii_type == "DATE_OF_BIRTH":
            replacement = "January 01, 1990"

        elif pii_type == "IP_ADDRESS":
            replacement = f"192.168.{rng.randint(1,254)}.{rng.randint(1,254)}"

        else:
            replacement = f"[REDACTED_{pii_type}]"

        self.mapping[clean_text] = replacement
        return replacement

File: js/faker_mapper.js
/**
 * Synthetic PII Replacement Engine in JavaScript
 * Ensures consistent mapping across document lifetime.
 */

class SyntheticPIIMapper {
    constructor(seed = 42) {
        self.seed = seed;
        this.mapping = {};

        // Pre-seeded explicit mappings matching assignment examples
        this.presets = {
            "Rashi Patil": "John Doe",
            "rashhi.patil@gmail.com": "john.doe@example.com",
            "Rohan Dey": "Peter Parker",
            "rohan.dey@gmail.com": "peter.parker@example.com",
            "+91 9876543210": "+91 1234567645"
        };
    }

    _simpleHash(str) {
        let hash = 0;
        for (let i = 0; i < str.length; i++) {
            const char = str.charCodeAt(i);
            hash = (hash << 5) - hash + char;
            hash |= 0; // Convert to 32bit integer
        }
        return Math.abs(hash);
    }

    getReplacement(originalText, piiType, mode = "FAKER") {
        const clean = originalText.trim();
        if (!clean) return originalText;

        if (mode === "MASK") {
            return `[REDACTED_${piiType}]`;
        }

        if (mode === "ASTERISK") {
            return clean.replace(/[a-zA-Z0-9]/g, "*");
        }

        // FAKER mode (consistent synthetic values)
        if (this.presets[clean]) {
            return this.presets[clean];
        }

        if (this.mapping[clean]) {
            return this.mapping[clean];
        }

        const hashVal = this._simpleHash(clean);
        let replacement = "";

        switch (piiType) {
            case "FULL_NAME": {
                const firsts = ["Alex", "Jordan", "Morgan", "Taylor", "Casey", "Riley", "Sam", "Chris", "David", "Sarah", "Michael", "Emily"];
                const lasts = ["Smith", "Johnson", "Williams", "Brown", "Jones", "Garcia", "Miller", "Davis", "Martinez"];
                const fIndex = hashVal % firsts.length;
                const lIndex = (hashVal >> 3) % lasts.length;
                replacement = `${firsts[fIndex]} ${lasts[lIndex]}`;
                break;
            }

            case "EMAIL": {
                const user = clean.split("@")[0].toLowerCase().replace(/[^a-z0-9.]/g, "");
                replacement = `${user}.fake@example.com`;
                break;
            }

            case "PHONE_NUMBER": {
                const numStr = (hashVal.toString() + "9876543210").slice(0, 10);
                if (clean.startsWith("+91")) {
                    replacement = `+91 ${numStr.slice(0, 5)} ${numStr.slice(5)}`;
                } else if (clean.startsWith("+")) {
                    replacement = `+1 555-${numStr.slice(0, 3)}-${numStr.slice(3, 7)}`;
                } else {
                    replacement = `+91 ${numStr}`;
                }
                break;
            }

            case "COMPANY_NAME": {
                const comps = ["Acme Corp", "Apex Technologies", "Global Industries", "Omni Enterprises", "Zenith Holdings", "Nexus Solutions"];
                replacement = comps[hashVal % comps.length];
                break;
            }

            case "ADDRESS": {
                const streets = ["123 Innovation Way", "456 Tech Park Blvd", "789 Corporate Plaza"];
                const cities = ["Metro City", "Techville", "Cyber City"];
                replacement = `${streets[hashVal % streets.length]}, ${cities[(hashVal >> 2) % cities.length]} - 400001, India`;
                break;
            }

            case "GOVT_ID": {
                if (clean.length === 10 && /^[A-Z]{5}/.test(clean)) {
                    replacement = "ABCDE" + (hashVal % 9000 + 1000) + "X";
                } else if (clean.length === 8) {
                    replacement = "0" + (hashVal % 9000007 + 1000000);
                } else {
                    replacement = "ID-" + (hashVal % 900000000 + 100000000);
                }
                break;
            }

            case "CREDIT_CARD": {
                replacement = "4111-XXXX-XXXX-1111";
                break;
            }

            case "DATE_OF_BIRTH": {
                replacement = "January 01, 1990";
                break;
            }

            case "IP_ADDRESS": {
                replacement = `192.168.${hashVal % 254 + 1}.${(hashVal >> 4) % 254 + 1}`;
                break;
            }

            default:
                replacement = `[REDACTED_${piiType}]`;
        }

        this.mapping[clean] = replacement;
        return replacement;
    }
}
